### PR TITLE
Feature/fast ids

### PR DESCRIPTION
**Pull request by andre-merzky**
Support for fast ID generation which provides process-local unique IDs w/o file access or system callouts

### CODE REVIEW

**Comment by mtitov** [comment]
@andre-merzky isn't it the same as `SIMPLE`? btw, it makes sense to update current approaches as `simple` -> `mem_counter` / `counter` (and it could be enhanced with fast-approach) and `item_counter` -> `file_counter`

**Comment by andre-merzky** [comment]
> @andre-merzky isn't it the same as `SIMPLE`? btw, it makes sense to update current approaches as `simple` -> `mem_counter` / `counter` (and it could be enhanced with fast-approach) and `item_counter` -> `file_counter`

I also wanted to avoid, most of all, the code paths we have in the 'proper' ID generation.  `FAST` is 4 times faster than `SIMPLE` mainly for avoiding those.  Adding a `mem_counter` and then using a single code path gives the same semantics, but not the same performance I'm afraid.

I stumbled over this when creating millions of tasks in a test.  `SIMPLE` performs ok (about 200k / second), but `FAST` moved that to about 900k/sec - that happened to remove a bottleneck for that test I did :-)

**Comment by mtitov** [comment]
Then I would propose to change the current approach:
- have special cases, e.g., `ID_UUID`, `ID_UNIQUE`, `ID_PRIVATE`, `ID_BASE`, with their specific methods (having a template for a name, but kinda not a template-approach)
- have a general case - `ID_CUSTOM` - with all that cases/options (aka template-approach - checking possible options)

Proposed `ID_FAST` is the same as `ID_SIMPLE`, but without lock mechanism (do we need that?) - maybe we can have just one of them as `ID_BASE`. Just it seems that bringing up `ID_FAST` this way breaks the consistency of the module

p.s., as I see, `ID_SIMPLE` is used in SAGA only (within RADICAL packages)

**Comment by andre-merzky** [comment]
So instead of one special case we have 5 special cases?  I don't think that improves things, really (and I also consider that refactoring out of scope for this PR).  Hmmm...

**Comment by mtitov** [comment]
> So instead of one special case we have 5 special cases? I don't think that improves things, really (and I also consider that refactoring out of scope for this PR).

Sorry, I disagree that it is out of scope.. while you introduce a new feature, you either follow the module design/approach, or you change it, so the new feature would fit (considering to continue support all existing cases). It seems that `ID_FAST` suppose to replace `ID_SIMPLE` (since why do we need the same functionality with two implementations?), but also the code is developed in a different way and I think it is reasonable to extend approach for a special case to other such cases (the general case - `custom` - suppose to reuse template parts from that special cases).

p.s. `ID_FASTEST` will be another new mode? ;)

(*) Have couple remarks to the current changes (will comment below), but my proposal is to rework this PR (as I wrote above)

File: src/radical/utils/ids.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 TEMPLATE_UNIQUE  = "%(prefix)s.%(date)s.%(time)s.%(pid)06d.%(counter)04d"
 TEMPLATE_PRIVATE = "%(prefix)s.%(host)s.%(user)s.%(days)06d.%(day_counter)04d"
 TEMPLATE_UUID    = "%(prefix)s.%(uuid)s"
+TEMPLATE_FAST    = "%(prefix)s.%(fast)06d"
```
Comment (mtitov) [inline src/radical/utils/ids.py:22]: should `TEMPLATE_SIMPLE` be reused? and update that template with `06d`?

**Comment by andre-merzky** [comment]
I don't think it's worth the effort.  Was just response to a perf test I did, so not important enough....

File: tests/unittests/test_ids.py [hunk 2]
```diff
@@ -55,14 +55,16 @@ def tearDownClass(cls):
...
     # --------------------------------------------------------------------------
     #
     def tearDown(self):
         ru.reset_id_counters(prefix=None, reset_all_others=True)
 
+
     # --------------------------------------------------------------------------
     #
-    def test_generate_id(self):
+    def test_generate_id(self, second=False):
```
Comment (mtitov) [inline tests/unittests/test_ids.py:67]: `second=False` - I assume that was used for local tests, and this is not needed here, right?

**Pull request closed without merging**